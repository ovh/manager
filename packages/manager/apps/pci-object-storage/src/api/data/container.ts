import { v6 } from '@ovh-ux/manager-core-api';

export type TObject = {
  key: string;
  name: string;
  etag: string;
  size: number;
  contentType: string;
  lastModified: string;
  storageClass: string;
  s3StorageType: string;
  retrievalState: string;
  retrievalDate: string;
};
export type TServerContainer = {
  createdAt: string;
  encryption: {
    sseAlgorithm: string;
  };
  id: string;
  name: string;
  objects: TObject[];
  objectsCount: number;
  objectsSize: number;
  ownerId: number;
  region: string;
  s3StorageType: unknown;
  storedBytes: number;
  storedObjects: number;
  versioning: {
    status: string;
  };
  virtualHost: string;
  staticUrl: string;
  replication: {
    rules: {
      status: string;
    }[];
  };
};

export const getServerContainer = async (
  projectId: string,
  region: string,
  name: string,
  id: string,
): Promise<TServerContainer> => {
  const url = id
    ? `/cloud/project/${projectId}/storage/${id}`
    : `/cloud/project/${projectId}/region/${region}/storage/${name}`;

  const { data } = await v6.get<TServerContainer>(url);
  return data;
};
