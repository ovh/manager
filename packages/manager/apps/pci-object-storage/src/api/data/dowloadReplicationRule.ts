import { saveAs } from 'file-saver';
import { TReplication } from '@/pages/objects/container/object/replication/ReplicationList.page';

export const downloadReplicationRules = (
  replications: TReplication[],
  projectId: string,
) => {
  const data = {
    Role: `arn:aws:iam::${projectId}:role/s3-replication`,
    Rules: replications,
  };

  const dataStr = JSON.stringify(data, null, 2);
  const blob = new Blob([dataStr], { type: 'application/json' });

  saveAs(blob, 'replicationRules.json');
};
