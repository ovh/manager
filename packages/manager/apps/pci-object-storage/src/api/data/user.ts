import { v6 } from '@ovh-ux/manager-core-api';

export type TS3Credentials = {
  userId: string;
  tenantId: string;
  access: string;
};

export type TUser = {
  id: number;
  username: string;
  creationDate: string;
  description: string;
  openstackId: string;
  status: string;
  access?: string;
  search?: string;
};

export const getAllUsers = async (projectId: string): Promise<TUser[]> => {
  const { data } = await v6.get<TUser[]>(`/cloud/project/${projectId}/user`);
  return data;
};

export const getS3Credentials = async (
  projectId: string,
  userId: string,
): Promise<TS3Credentials[]> => {
  const { data } = await v6.get<TS3Credentials[]>(
    `/cloud/project/${projectId}/user/${userId}/s3Credentials`,
  );
  return data;
};

export const getUser = async (
  projectId: string,
  userId: string,
): Promise<TUser> => {
  const { data } = await v6.get<TUser>(
    `/cloud/project/${projectId}/user/${userId}`,
  );
  return data;
};

export const deleteUser = async (
  projectId: string,
  userId: string,
  accessKey: string,
) => {
  await v6.delete(
    `/cloud/project/${projectId}/user/${userId}/s3Credentials/${accessKey}`,
  );
};
