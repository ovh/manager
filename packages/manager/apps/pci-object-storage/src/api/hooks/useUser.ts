import { useMutation, useQueries, useQuery } from '@tanstack/react-query';
import { ColumnSort, PaginationState } from '@ovh-ux/manager-react-components';
import { applyFilters, Filter } from '@ovh-ux/manager-core-api';
import { useMemo } from 'react';
import { paginateResults, sortResults } from '@/helpers';
import {
  deleteUser,
  getAllUsers,
  getS3Credentials,
  TS3Credentials,
  TUser,
} from '@/api/data/user';
import queryClient from '@/queryClient';

const getQueryKeyUsers = (projectId: string) => ['project', projectId, 'users'];

export const useAllUsers = (projectId: string) =>
  useQuery({
    queryKey: getQueryKeyUsers(projectId),
    queryFn: () => getAllUsers(projectId),
  });

export const useUsers = (projectId: string) => {
  const { data: users, isPending } = useAllUsers(projectId);
  return useQueries({
    queries: (users || [])?.map((user) => ({
      queryKey: [...getQueryKeyUsers(projectId), user.id, 's3Credentials'],
      queryFn: () => getS3Credentials(projectId, `${user.id}`),
      enabled: !isPending,
      select: (s3Credentials: TS3Credentials[]) => s3Credentials[0],
    })),
    combine: (results) => ({
      isPending: results.some((result) => result.isPending),
      isLoading: results.some((result) => result.isLoading),
      error: results.find((result) => result.error),
      data: users?.reduce<TUser[]>((all, user) => {
        const s3Credentials = results.find(
          (result) => result.data?.userId === user.openstackId,
        )?.data;

        if (s3Credentials) {
          all.push({
            ...user,
            access: s3Credentials.access,
            search: `${user.username} ${user.description} ${s3Credentials.access}`,
          });
        }

        return all;
      }, []),
    }),
  });
};

export const usePaginatedUsers = (
  projectId: string,
  pagination: PaginationState,
  sorting: ColumnSort,
  filters: Filter[],
) => {
  const { data: users, error, isLoading, isPending } = useUsers(projectId);

  return useMemo(
    () => ({
      isLoading,
      isPending,
      paginatedUsers: paginateResults<TUser>(
        sortResults<TUser>(applyFilters<TUser>(users || [], filters), sorting),
        pagination,
      ),
      allUsers: users,
      error,
    }),
    [users, error, isLoading, isPending, pagination, sorting, filters],
  );
};

type DeleteUserProps = {
  projectId: string;
  userId: string;
  access: string;
  onError: (cause: Error) => void;
  onSuccess: () => void;
};

export const useDeleteUser = ({
  projectId,
  userId,
  access,
  onError,
  onSuccess,
}: DeleteUserProps) => {
  const mutation = useMutation({
    mutationFn: async () => deleteUser(projectId, userId, access),
    onError,
    onSuccess: async () => {
      await queryClient.invalidateQueries({
        queryKey: ['project', projectId, 'users'],
      });
      onSuccess();
    },
  });
  return {
    deleteUser: () => mutation.mutate(),
    ...mutation,
  };
};
