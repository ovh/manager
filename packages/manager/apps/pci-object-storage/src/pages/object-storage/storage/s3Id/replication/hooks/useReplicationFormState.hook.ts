import { UseFormReturn } from 'react-hook-form';
import { ReplicationFormValues } from '../schemas/replicationForm.schema';

export const useReplicationFormState = (
  form: UseFormReturn<ReplicationFormValues>,
) => {
  const formValues = form.watch();

  const hasTags =
    formValues.tags &&
    Object.values(formValues.tags).some(
      (tag) => tag.key.trim() !== '' || tag.value.trim() !== '',
    );
  const isDeleteMarkerEnabled =
    formValues.deleteMarkerReplication === 'enabled';

  return {
    showScopeFields: formValues.isReplicationApplicationLimited,
    showStorageClassField: formValues.useStorageClass,
    isDeleteMarkerDisabled: hasTags,
    isTagsDisabled: isDeleteMarkerEnabled,
  };
};
