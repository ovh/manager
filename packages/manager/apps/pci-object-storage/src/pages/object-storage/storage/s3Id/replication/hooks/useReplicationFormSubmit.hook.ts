import { AddReplicationFormValues } from '../new/formAddReplication/useAddReplicationForm.hook';
import { transformTagsToApi } from '../utils/transformTags.util';
import storages, { FormattedStorage } from '@/types/Storages';

interface UseReplicationFormSubmitParams {
  availableDestinations: FormattedStorage[];
  s3Region?: string;
  s3Name?: string;
}

export const useReplicationFormSubmit = ({
  availableDestinations,
  s3Region,
  s3Name,
}: UseReplicationFormSubmitParams) => {
  const buildReplicationRule = (
    formValues: AddReplicationFormValues,
  ): storages.ReplicationRule | null => {
    const selectedDestination = availableDestinations.find(
      (storage) =>
        `${storage.name}:${storage.region}` === formValues.destination,
    );

    if (!selectedDestination || !s3Region || !s3Name) return null;

    const {
      tags,
      prefix,
      isReplicationApplicationLimited,
      status,
      ruleId,
      useStorageClass,
      storageClass,
      deleteMarkerReplication,
      priority,
    } = formValues;

    const transformedTags = transformTagsToApi(tags);

    const hasPrefix = isReplicationApplicationLimited && prefix;
    const hasTags =
      isReplicationApplicationLimited &&
      Object.keys(transformedTags).length > 0;
    const hasFilter = hasPrefix || hasTags;

    return {
      id: ruleId,
      status,
      destination: {
        name: selectedDestination.name,
        region: selectedDestination.region,
        ...(useStorageClass && {
          storageClass,
        }),
      },
      deleteMarkerReplication,
      priority: parseInt(priority, 10),
      ...(hasFilter && {
        filter: {
          ...(hasPrefix && { prefix }),
          ...(hasTags && { tags: transformedTags }),
        },
      }),
    };
  };

  return { buildReplicationRule };
};
