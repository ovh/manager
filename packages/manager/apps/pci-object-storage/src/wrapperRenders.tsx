import { QueryClient, QueryClientProvider } from '@tanstack/react-query';
import {
  ShellContext,
  ShellContextType,
} from '@ovh-ux/manager-react-shell-client';
import { vi } from 'vitest';
import { createMemoryRouter, RouterProvider } from 'react-router-dom';
import React from 'react';

export const shellContext = {
  environment: {
    getUser: () => ({ ovhSubsidiary: 'mocked_ovhSubsidiary' }),
  },
  shell: {
    navigation: {
      getURL: vi.fn(),
    },
    tracking: {
      trackPage: vi.fn(),
      trackClick: vi.fn(),
    },
  },
};

const queryClient = new QueryClient({
  defaultOptions: {
    mutations: {
      retry: false,
    },
    queries: {
      retry: false,
    },
  },
});

export const wrapper = ({ children }) => (
  <QueryClientProvider client={queryClient}>
    <ShellContext.Provider
      value={(shellContext as unknown) as ShellContextType}
    >
      {children}
    </ShellContext.Provider>
  </QueryClientProvider>
);

export const mockShellContext = {
  environment: {
    getRegion: vi.fn().mockReturnValue('EU'),
    getUser: vi.fn().mockReturnValue({ ovhSubsidiary: 'mocked_ovhSubsidiary' }),
    getApplication: vi.fn(),
  },
  shell: {
    environment: {
      getRegion: vi.fn().mockReturnValue('EU'),
      getUser: vi.fn().mockReturnValue({}),
      getApplication: vi.fn(),
    },
    navigation: {
      getURL: vi.fn(),
    },
    tracking: {
      trackPage: vi.fn(),
      trackClick: vi.fn(),
    },
  },
};

export const wrapperShow = ({ children }: { children: React.ReactNode }) => {
  const router = createMemoryRouter(
    [
      {
        path: '/',
        element: children,
        handle: {
          crumb: () => 'Home',
        },
      },
    ],
    {
      initialEntries: ['/'],
      future: {
        v7_startTransition: true,
        v7_relativeSplatPath: true,
      } as any,
    },
  );

  return (
    <QueryClientProvider client={queryClient}>
      <ShellContext.Provider value={mockShellContext as any}>
        <RouterProvider router={router} />
      </ShellContext.Provider>
    </QueryClientProvider>
  );
};

export const wrapperOffsiteReplication = ({
  children,
}: {
  children: React.ReactNode;
}) => (
  <QueryClientProvider client={queryClient}>
    <ShellContext.Provider value={mockShellContext as any}>
      {children}
    </ShellContext.Provider>
  </QueryClientProvider>
);
