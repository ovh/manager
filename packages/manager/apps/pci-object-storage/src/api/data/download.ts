import { v6 } from '@ovh-ux/manager-core-api';
import { addWeeks } from 'date-fns';
import { OPENIO_PRESIGN_EXPIRE } from '@/constants';
import { TContainer } from '@/pages/objects/container/object/show/Show.page';
import { TObject } from '@/api/data/container';

export const downloadObject = async (
  object: TObject,
  container: TContainer,
  projectId: string,
) => {
  let downloadUrl;

  if (container.s3StorageType) {
    const presignUrl = `/cloud/project/${projectId}/region/${container.region}/${container.s3StorageType}/${container.name}/presign`;
    const presignOptions = {
      expire: OPENIO_PRESIGN_EXPIRE,
      method: 'GET',
      object: object.key,
      ...(object.versionId && { versionId: object.versionId }),
    };

    const { data: presignData } = await v6.post<{ url: string }>(
      presignUrl,
      presignOptions,
    );
    downloadUrl = presignData.url;
  } else {
    const expirationDate = addWeeks(new Date(), 1).toISOString();
    const publicUrl = `/cloud/project/${projectId}/storage/${container.id}/publicUrl`;
    const publicOptions = {
      expirationDate,
      objectName: object.name,
    };

    const { data: publicData } = await v6.post<{ getURL: string }>(
      publicUrl,
      publicOptions,
    );
    downloadUrl = publicData.getURL;
  }

  const response = await fetch(downloadUrl);
  const blob = await response.blob();

  const blobUrl = window.URL.createObjectURL(blob);
  const link = document.createElement('a');
  link.href = blobUrl;
  link.download = object.key || object.name;
  document.body.appendChild(link);
  link.click();

  window.URL.revokeObjectURL(blobUrl);
  document.body.removeChild(link);
};
