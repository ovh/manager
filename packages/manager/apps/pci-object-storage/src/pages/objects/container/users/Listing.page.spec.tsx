import { describe, vi } from 'vitest';
import { render } from '@testing-library/react';
import { useFeatureAvailability } from '@ovh-ux/manager-react-components';
import { usePaginatedUsers, usePostS3Secret } from '@/api/hooks/useUser';
import ListingPage from './Listing.page';
import { wrapper } from '@/wrapperRenders';
import { TUser } from '@/api/data/user';

vi.mock('@/api/hooks/useUser', () => {
  return {
    usePaginatedUsers: vi.fn(),
    usePostS3Secret: vi.fn(),
  };
});

describe('ListingPage', () => {
  it('sould display spinner', () => {
    vi.mocked(usePaginatedUsers).mockReturnValue({
      isPending: true,
      paginatedUsers: [],
    } as never);
    vi.mocked(useFeatureAvailability).mockReturnValue({
      isPending: false,
    } as never);
    const { asFragment } = render(<ListingPage />, { wrapper });
    expect(asFragment()).toMatchSnapshot();
  });

  it('sould display datagrid', () => {
    vi.mocked(usePostS3Secret).mockReturnValue({
      postS3Secret: vi.fn(),
    } as never);
    const paginatedUsers = {
      rows: ([
        {
          id: 'id',
          username: 'username',
          description: 'description',
          status: 'status',
          password: 'password',
        },
      ] as unknown) as TUser[],
    };
    vi.mocked(usePaginatedUsers).mockReturnValue({
      isPending: false,
      paginatedUsers,
    } as never);
    vi.mocked(useFeatureAvailability).mockReturnValue({
      data: [{ feature: false }],
      isPending: false,
    } as never);
    const { container } = render(<ListingPage />, { wrapper });
    expect(container).toMatchSnapshot();
  });
});
