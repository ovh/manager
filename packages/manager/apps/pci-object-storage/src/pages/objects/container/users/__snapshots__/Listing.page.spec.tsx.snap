// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ListingPage > sould display datagrid 1`] = `
<div>
  <div
    class="header mt-8"
  >
    <div
      data-testid="notifications"
    />
  </div>
  <ods-text
    class="mt-6 block"
    preset="heading-4"
  >
    pci_projects_project_storages_containers_users_title
  </ods-text>
  <ods-text
    class="mt-6 block"
    preset="paragraph"
  >
    pci_projects_project_storages_containers_users_user_description
  </ods-text>
  <div
    class="sm:flex items-center justify-between mt-8"
  >
    <ods-button
      class="xs:mb-0.5 sm:mb-0"
      color="primary"
      icon="plus"
      label="pci_projects_project_storages_containers_users_add_user"
      size="sm"
    />
    <div
      class="flex justify-center gap-4"
    >
      <ods-button
        class="xs:mb-0.5 sm:mb-0"
        color="primary"
        icon="refresh"
        label=""
        size="sm"
        variant="outline"
      />
      <ods-input
        class="min-w-[15rem]"
        name="searchField"
        value=""
      />
      <ods-button
        icon="magnifying-glass"
        label=""
        size="sm"
      />
      <ods-button
        color="primary"
        icon="filter"
        id="popover-filter"
        label="pci-common:common_criteria_adder_filter_label"
        size="sm"
        slot="popover-trigger"
        variant="outline"
      />
      <ods-popover
        trigger-id="popover-filter"
      >
        <div>
          <ods-form-field
            class="w-full"
          >
            <div
              slot="label"
            >
              <span
                class="text-[--ods-color-heading] leading-[22px]"
              >
                common_criteria_adder_column_label
              </span>
            </div>
            <ods-select
              data-testid="add-filter_select_idColumn"
              name="add-filter_select_idColumn"
              value="username"
            >
              <option
                value="username"
              >
                pci_projects_project_storages_containers_users_username
              </option>
              <option
                value="description"
              >
                pci_projects_project_storages_containers_users_description
              </option>
              <option
                value="access"
              >
                pci_projects_project_storages_containers_users_accesskey
              </option>
            </ods-select>
          </ods-form-field>
        </div>
        <div>
          <ods-form-field
            class="mt-2 w-full"
          >
            <div
              slot="label"
            >
              <span
                class="text-[--ods-color-heading] leading-[22px]"
              >
                common_criteria_adder_operator_label
              </span>
            </div>
            <div>
              <ods-select
                name="add-operator-username"
                value="includes"
              >
                <option
                  value="includes"
                >
                  common_criteria_adder_operator_includes
                </option>
                <option
                  value="starts_with"
                >
                  common_criteria_adder_operator_starts_with
                </option>
                <option
                  value="ends_with"
                >
                  common_criteria_adder_operator_ends_with
                </option>
                <option
                  value="is_equal"
                >
                  common_criteria_adder_operator_is_equal
                </option>
                <option
                  value="is_different"
                >
                  common_criteria_adder_operator_is_different
                </option>
              </ods-select>
            </div>
          </ods-form-field>
        </div>
        <div>
          <ods-form-field
            class="mt-2 w-full"
          >
            <div
              slot="label"
            >
              <span
                class="text-[--ods-color-heading] leading-[22px]"
              >
                common_criteria_adder_value_label
              </span>
            </div>
            <ods-input
              class="border"
              data-testid="filter-add_value-input"
              name="filter-add_value-input"
              type="text"
              value=""
            />
          </ods-form-field>
        </div>
        <div>
          <ods-button
            class="mt-4 w-full filter-add-button-submit"
            data-testid="filter-add_submit"
            is-disabled="true"
            label="common_criteria_adder_submit_label"
            size="sm"
          />
        </div>
      </ods-popover>
    </div>
  </div>
  <div
    class="my-8 mt-8"
  />
  <div
    class="mt-8"
  >
    <div>
      <div
        class="contents px-[1px] overflow-x-visible"
      >
        <ods-table
          class="overflow-x-visible"
        >
          <!---->
          <table
            class="w-full border-collapse"
            style="--expander-column-width: 2.5rem;"
          >
            <thead>
              <tr>
                <th
                  class="text-left pl-4 h-11 whitespace-nowrap "
                >
                  <div
                    class="cursor-pointer select-none"
                    data-testid="header-username"
                  >
                    <span>
                      pci_projects_project_storages_containers_users_username
                    </span>
                    <span
                      class="align-middle inline-block h-4 -mt-6"
                    >
                      <ods-icon
                        class="invisible"
                        name="arrow-down"
                      />
                    </span>
                  </div>
                </th>
                <th
                  class="text-left pl-4 h-11 whitespace-nowrap "
                >
                  <div
                    class="cursor-pointer select-none"
                    data-testid="header-description"
                  >
                    <span>
                      pci_projects_project_storages_containers_users_description
                    </span>
                    <span
                      class="align-middle inline-block h-4 -mt-6"
                    >
                      <ods-icon
                        class="invisible"
                        name="arrow-down"
                      />
                    </span>
                  </div>
                </th>
                <th
                  class="text-left pl-4 h-11 whitespace-nowrap "
                >
                  <div
                    class="cursor-pointer select-none"
                    data-testid="header-access"
                  >
                    <span>
                      pci_projects_project_storages_containers_users_accesskey
                    </span>
                    <span
                      class="align-middle inline-block h-4 -mt-6"
                    >
                      <ods-icon
                        class="invisible"
                        name="arrow-down"
                      />
                    </span>
                  </div>
                </th>
                <th
                  class="text-left pl-4 h-11 whitespace-nowrap "
                >
                  <div
                    class=""
                    data-testid="header-actions"
                  >
                    <span />
                    <span
                      class="align-middle inline-block h-4 -mt-6"
                    >
                      <ods-icon
                        class="invisible"
                        name="arrow-down"
                      />
                    </span>
                  </div>
                </th>
              </tr>
            </thead>
            <tbody>
              <tr
                class="border-solid border-[1px] h-[3.25rem] border-[--ods-color-blue-200]"
              >
                <td
                  class="text-left pl-4"
                >
                  <ods-text
                    preset="span"
                  >
                    username
                  </ods-text>
                </td>
                <td
                  class="text-left pl-4"
                >
                  <ods-text
                    preset="span"
                  >
                    description
                  </ods-text>
                </td>
                <td
                  class="text-left pl-4"
                >
                  <ods-text
                    preset="span"
                  >
                    common_empty_text_cell
                  </ods-text>
                </td>
                <td
                  class="text-left pl-4"
                >
                  <div
                    class="min-w-16"
                  >
                    <div
                      class="w-min"
                      id="navigation-action-trigger-id"
                    >
                      <ods-button
                        data-testid="navigation-action-trigger-action"
                        icon="ellipsis-vertical"
                        id="id"
                        is-disabled="false"
                        is-loading="false"
                        size="sm"
                        slot="menu-title"
                        variant="outline"
                      />
                    </div>
                    <ods-popover
                      class="py-[8px] px-0 w-max"
                      trigger-id="navigation-action-trigger-id"
                      with-arrow="true"
                    >
                      <div
                        class="flex flex-col"
                      >
                        <ods-button
                          class="menu-item-button w-full"
                          display-tooltip="false"
                          label="pci_projects_project_storages_containers_users_import_json"
                          size="sm"
                          variant="ghost"
                        />
                        <ods-button
                          class="menu-item-button w-full"
                          display-tooltip="false"
                          label="pci_projects_project_storages_containers_users_download_json"
                          size="sm"
                          variant="ghost"
                        />
                        <ods-button
                          class="menu-item-button w-full"
                          display-tooltip="false"
                          label="pci_projects_project_storages_containers_users_download_rclone_file"
                          size="sm"
                          variant="ghost"
                        />
                        <ods-button
                          class="menu-item-button w-full"
                          display-tooltip="false"
                          label="pci_projects_project_storages_containers_users_see_secret_key"
                          size="sm"
                          variant="ghost"
                        />
                        <ods-button
                          class="menu-item-button w-full"
                          display-tooltip="false"
                          label="pci_projects_project_storages_containers_delete_label"
                          size="sm"
                          variant="ghost"
                        />
                      </div>
                    </ods-popover>
                  </div>
                </td>
              </tr>
            </tbody>
          </table>
        </ods-table>
      </div>
      <ods-pagination
        class="flex xs:justify-start md:justify-end my-8"
        default-current-page="1"
        default-items-per-page="10"
        total-items="0"
        total-pages="0"
      >
        <span
          class="mr-3"
          slot="before-total-items"
        >
          common_pagination_of
        </span>
        <span
          class="ml-3"
          slot="after-total-items"
        >
          common_pagination_results
        </span>
      </ods-pagination>
    </div>
  </div>
  Outlet
</div>
`;

exports[`ListingPage > sould display spinner 1`] = `
<DocumentFragment>
  <div
    class="header mt-8"
  >
    <div
      data-testid="notifications"
    />
  </div>
  <ods-text
    class="mt-6 block"
    preset="heading-4"
  >
    pci_projects_project_storages_containers_users_title
  </ods-text>
  <ods-text
    class="mt-6 block"
    preset="paragraph"
  >
    pci_projects_project_storages_containers_users_user_description
  </ods-text>
  <div
    class="sm:flex items-center justify-between mt-8"
  >
    <ods-button
      class="xs:mb-0.5 sm:mb-0"
      color="primary"
      icon="plus"
      label="pci_projects_project_storages_containers_users_add_user"
      size="sm"
    />
    <div
      class="flex justify-center gap-4"
    >
      <ods-button
        class="xs:mb-0.5 sm:mb-0"
        color="primary"
        icon="refresh"
        label=""
        size="sm"
        variant="outline"
      />
      <ods-input
        class="min-w-[15rem]"
        name="searchField"
        value=""
      />
      <ods-button
        icon="magnifying-glass"
        label=""
        size="sm"
      />
      <ods-button
        color="primary"
        icon="filter"
        id="popover-filter"
        label="pci-common:common_criteria_adder_filter_label"
        size="sm"
        slot="popover-trigger"
        variant="outline"
      />
      <ods-popover
        trigger-id="popover-filter"
      >
        <div>
          <ods-form-field
            class="w-full"
          >
            <div
              slot="label"
            >
              <span
                class="text-[--ods-color-heading] leading-[22px]"
              >
                common_criteria_adder_column_label
              </span>
            </div>
            <ods-select
              data-testid="add-filter_select_idColumn"
              name="add-filter_select_idColumn"
              value="username"
            >
              <option
                value="username"
              >
                pci_projects_project_storages_containers_users_username
              </option>
              <option
                value="description"
              >
                pci_projects_project_storages_containers_users_description
              </option>
              <option
                value="access"
              >
                pci_projects_project_storages_containers_users_accesskey
              </option>
            </ods-select>
          </ods-form-field>
        </div>
        <div>
          <ods-form-field
            class="mt-2 w-full"
          >
            <div
              slot="label"
            >
              <span
                class="text-[--ods-color-heading] leading-[22px]"
              >
                common_criteria_adder_operator_label
              </span>
            </div>
            <div>
              <ods-select
                name="add-operator-username"
                value="includes"
              >
                <option
                  value="includes"
                >
                  common_criteria_adder_operator_includes
                </option>
                <option
                  value="starts_with"
                >
                  common_criteria_adder_operator_starts_with
                </option>
                <option
                  value="ends_with"
                >
                  common_criteria_adder_operator_ends_with
                </option>
                <option
                  value="is_equal"
                >
                  common_criteria_adder_operator_is_equal
                </option>
                <option
                  value="is_different"
                >
                  common_criteria_adder_operator_is_different
                </option>
              </ods-select>
            </div>
          </ods-form-field>
        </div>
        <div>
          <ods-form-field
            class="mt-2 w-full"
          >
            <div
              slot="label"
            >
              <span
                class="text-[--ods-color-heading] leading-[22px]"
              >
                common_criteria_adder_value_label
              </span>
            </div>
            <ods-input
              class="border"
              data-testid="filter-add_value-input"
              name="filter-add_value-input"
              type="text"
              value=""
            />
          </ods-form-field>
        </div>
        <div>
          <ods-button
            class="mt-4 w-full filter-add-button-submit"
            data-testid="filter-add_submit"
            is-disabled="true"
            label="common_criteria_adder_submit_label"
            size="sm"
          />
        </div>
      </ods-popover>
    </div>
  </div>
  <div
    class="my-8 mt-8"
  />
  <ods-spinner
    class="mt-8"
    data-testid="List-spinner"
    size="md"
  />
  Outlet
</DocumentFragment>
`;
